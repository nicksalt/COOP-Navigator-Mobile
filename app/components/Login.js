import React from "react";
import {
  ActivityIndicator,
  TouchableOpacity,
  Text,
  TextInput,
  View,
  KeyboardAvoidingView,
  StyleSheet,
  Image,
  Switch
} from "react-native";
import {
  saveLogin,
  getAutoLogin,
  setAutoLogin,
  getLoginInfo,
  forgetLogin
} from "../config/auth";

export default class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      rememberMe: true,
      email: "amizo067@uottawa.ca",
      password: "lethalpassword"
    };
  }

  componentDidMount = () => {
    this.bootstrapLogin();
  };

  componentDidUpdate = async (prevProps, prevState) => {
    const { rememberMe, email, password } = this.state;
    if (!prevProps.isLoggedIn && this.props.isLoggedIn) {
      if (rememberMe) {
        setAutoLogin(rememberMe);
        saveLogin(email, password);
      }
      this.props.navigation.navigate("AppScreens");
    }
    if (this.props.loginError) {
      await setAutoLogin(false);
      await forgetLogin();
    }
  };

  bootstrapLogin = async () => {
    try {
      console.log(
        "This print should be followed by one saying AUTOLOGIN SUCCESS." +
          "Remote debugging cases errors with AsyncStorage"
      );
      const autoLogin = await getAutoLogin();
      console.log("AUTOLOGIN SUCCESS", autoLogin);
      if (autoLogin) {
        const { email, password } = await getLoginInfo();
        if (email && password) {
          this.props.loginDispatch(email, password);
        }
      }
    } catch (err) {
      console.error(err);
    }
  };

  onLoginClick = async () => {
    try {
      let { email, password } = this.state;
      email = email.toLowerCase();
      this.props.loginDispatch(email, password);
    } catch (err) {
      console.error("onLoginClick Error", err);
    }
  };

  displayError() {
    let message = " ";
    if (this.props.loginError) message = this.props.loginError.message;
    return <Text style={styles.errorMessage}>{message}</Text>;
  }

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.logoContainer}>
          <Image
            style={styles.logo}
            source={require("../images/uottawa_Logo.png")}
          />
          <Text style={styles.subtitle}>Co-op</Text>
        </View>

        <KeyboardAvoidingView style={styles.loginContainer} behavior="padding">
          <TextInput
            style={styles.input}
            placeholder="Email"
            underlineColorAndroid="transparent"
            value={this.state.email}
            onChangeText={email => this.setState({ email })}
          />

          <TextInput
            secureTextEntry={true}
            style={styles.input}
            placeholder="Password"
            underlineColorAndroid="transparent"
            value={this.state.password}
            onChangeText={password => this.setState({ password })}
          />
          <View style={styles.rememberMeContainer}>
            <Switch
              style={styles.rememberMeSwitch}
              title="Remember Me"
              value={this.state.rememberMe}
              onValueChange={() =>
                this.setState({ rememberMe: !this.state.rememberMe })
              }
            />
            <Text style={styles.rememberMeText}>Remember Me</Text>
          </View>
          <TouchableOpacity
            style={styles.buttonContainer}
            onPress={this.onLoginClick}
            title="LOGIN"
          >
            <Text style={styles.loginButton}>LOGIN </Text>
            {this.props.attemptingLogin && (
              <View style={styles.activityIndicatorContainer}>
                <ActivityIndicator size="small" color="gray" />
              </View>
            )}
          </TouchableOpacity>

          {this.displayError()}
        </KeyboardAvoidingView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  activityIndicatorContainer: {
    marginLeft: 4
  },
  container: {
    flex: 1,
    justifyContent: "center"
  },
  logoContainer: {
    alignItems: "center",
    marginBottom: 20
  },
  logo: {
    width: 250,
    height: 215,
    marginBottom: 20
  },
  subtitle: {
    fontSize: 30,
    fontWeight: "500"
  },
  loginContainer: {
    padding: 20
  },
  input: {
    height: 50,
    backgroundColor: "rgba(255,255,255,0.7)",
    paddingHorizontal: 20,
    borderRadius: 15,
    marginBottom: 10
  },
  buttonContainer: {
    backgroundColor: "#a9343a",
    padding: 15,
    borderRadius: 15,
    marginTop: 20,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center"
  },
  loginButton: {
    textAlign: "center",
    fontWeight: "600",
    color: "white"
  },
  errorMessage: {
    textAlign: "center",
    color: "red"
  },
  rememberMeContainer: {
    flexDirection: "row",
    alignItems: "center",
    marginTop: "2%"
  },
  rememberMeSwitch: {
    marginRight: "3%"
  },
  rememberMeText: {
    fontSize: 16
  }
});
