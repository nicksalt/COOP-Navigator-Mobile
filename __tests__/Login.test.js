import React from "react";
import Login from "../app/components/Login";
import renderer from "react-test-renderer";

//login test
test("Login snapshot renders correctly", () => {
  const props = {};
  let tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
  props.attemptingLogin = true;
  tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
  props.loginError = { message: "FAILED LOGIN" };
  tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
});
