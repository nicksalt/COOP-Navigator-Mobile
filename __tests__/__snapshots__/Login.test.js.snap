// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginBottom": 20,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 215,
          "marginBottom": 20,
          "width": 250,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 30,
          "fontWeight": "500",
        }
      }
    >
      Co-op
    </Text>
  </View>
  <View
    enabled={true}
    keyboardVerticalOffset={0}
    onLayout={[Function]}
    style={
      Array [
        Object {
          "padding": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Email"
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 15,
          "height": 50,
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value="amizo067@uottawa.ca"
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 15,
          "height": 50,
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value="lethalpassword"
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": "2%",
        }
      }
    >
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {
              "marginRight": "3%",
            },
          ]
        }
        title="Remember Me"
        value={true}
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
          }
        }
      >
        Remember Me
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#a9343a",
          "borderRadius": 15,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontWeight": "600",
            "textAlign": "center",
          }
        }
      >
        LOGIN 
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "red",
          "textAlign": "center",
        }
      }
    >
       
    </Text>
  </View>
</View>
`;

exports[`Login snapshot renders correctly 2`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginBottom": 20,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 215,
          "marginBottom": 20,
          "width": 250,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 30,
          "fontWeight": "500",
        }
      }
    >
      Co-op
    </Text>
  </View>
  <View
    enabled={true}
    keyboardVerticalOffset={0}
    onLayout={[Function]}
    style={
      Array [
        Object {
          "padding": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Email"
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 15,
          "height": 50,
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value="amizo067@uottawa.ca"
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 15,
          "height": 50,
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value="lethalpassword"
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": "2%",
        }
      }
    >
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {
              "marginRight": "3%",
            },
          ]
        }
        title="Remember Me"
        value={true}
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
          }
        }
      >
        Remember Me
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#a9343a",
          "borderRadius": 15,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontWeight": "600",
            "textAlign": "center",
          }
        }
      >
        LOGIN 
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "red",
          "textAlign": "center",
        }
      }
    >
       
    </Text>
  </View>
</View>
`;

exports[`Login snapshot renders correctly 3`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginBottom": 20,
      }
    }
  >
    <Image
      source={1}
      style={
        Object {
          "height": 215,
          "marginBottom": 20,
          "width": 250,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 30,
          "fontWeight": "500",
        }
      }
    >
      Co-op
    </Text>
  </View>
  <View
    enabled={true}
    keyboardVerticalOffset={0}
    onLayout={[Function]}
    style={
      Array [
        Object {
          "padding": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Email"
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 15,
          "height": 50,
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value="amizo067@uottawa.ca"
    />
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,0.7)",
          "borderRadius": 15,
          "height": 50,
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
      underlineColorAndroid="transparent"
      value="lethalpassword"
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": "2%",
        }
      }
    >
      <RCTSwitch
        disabled={false}
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "height": 31,
              "width": 51,
            },
            Object {
              "marginRight": "3%",
            },
          ]
        }
        title="Remember Me"
        value={true}
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 16,
          }
        }
      >
        Remember Me
      </Text>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#a9343a",
          "borderRadius": 15,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginTop": 20,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontWeight": "600",
            "textAlign": "center",
          }
        }
      >
        LOGIN 
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "red",
          "textAlign": "center",
        }
      }
    >
       
    </Text>
  </View>
</View>
`;
